YALB-1198: Events: Exposed filters on dynamic list (BE)

codechefmarc opened the pull request:
## [YALB-1198: Events: Exposed filters on dynamic list (BE)](https://yaleits.atlassian.net/browse/YALB-1198)

### Description of work
- Adds event category taxonomy vocabulary
- Adds event category as a term reference into events
- Adds event category exposed filter to calendar list - limits terms to only ones in use
- Adds event category filtering to views basic - visible only when event is selected
- Adds operator (AND/OR) functionality to views basic
- Allows multiple terms to be selected / filtered on for tags and event categories (hence why operator was added)
- Adds "No results found." message to view if no results are found.

### Functional testing steps:
- [ ] Visit `/admin/structure/taxonomy/manage/event_category/add` and `/admin/structure/taxonomy/manage/tags/overview` and add a few terms to each vocabulary.
- [ ] Add multiple events to the site. Make sure to add dates in the future or else they won't show up in views.
- [x] Verify there is a new "Category" taxonomy field on the event edit form.
- [x] Verify the new autocomplete deluxe field allows to select existing and create new terms - this works best with some terms already in the system.

![Screenshot 2023-05-17 at 3 47 12 PM](https://github.com/yalesites-org/yalesites-project/assets/107938318/ef178dc2-34f8-47e5-a8d1-5b56d2c9b7de)

- [x] For each event, tag and categorize with some events getting the same terms and others getting different terms. For testing, use the term in the title to help with filtering later.
- [x] Create a new page, title it, save, and enter layout builder.
- [x] Add a block of type "Calendar List".  After adding the block, save layout builder.
- [x] Verify that the calendar list has an "Event Category" filter on the top. Note: This is currently un-styled.

![Screenshot 2023-05-17 at 3 52 15 PM](https://github.com/yalesites-org/yalesites-project/assets/107938318/ff573f58-5c06-4fd1-b5de-8770a0109b63)

- [ ] Verify that the terms listed in the filter are the only ones that are currently in use - it does not show all terms.
- [x] Re-enter layout builder and add a block of type "View"
- [x] Verify that there is a new "Match Content That Has" area with two options that control how terms should be matched to nodes (Any = OR, All = AND)

![Screenshot 2023-05-17 at 4 00 18 PM](https://github.com/yalesites-org/yalesites-project/assets/107938318/13a968be-c57c-41ce-8d4e-8a369d28c48c)

- [x] Select "Events" under "I Want To Show"
- [x] Verify that the "Event Category" autocomplete text box is added to the form

![Screenshot 2023-05-17 at 4 00 23 PM](https://github.com/yalesites-org/yalesites-project/assets/107938318/6ae87dee-bc23-426a-bd03-81242e7ef29d)

- [x] Add in some tags / categories via autocomplete to filter by and select either the "Any" or the "All" under the "Match Content That Has". Note, tags and categories can support multiple terms now.
- [x] Add the block.
- [x] Verify that there are results for the filters (Can be done without saving layout builder)
- [x] If there are filters that don't match any content, verify that there is a "No results found." message that appears for that block.
- [x] Configure the view block and change settings to check to see that the filters all work - i.e. test multiple terms, both tags and categories, and different "Any/All" combos to verify that they work as expected.

dblanken-yale commented:
@codechefmarc In the first step, I'm unable to visit the first link `event_category/add`; it says I don't have access with my CAS login--is that intended for our non-admin access level?  I can get to the overview.

dblanken-yale commented:
Great work.  I finished testing the rest then.  Really liking the filtering.  Not sure if you want to wait for Jim but looks good to me.  :D

codechefmarc commented:
@joetower - OK, I was able to theme the filter form and would like your eyes on the code, specifically in atomic and component library. Here's what I did:
- CSS is in `01-atoms/forms/_yds-form.scss` (Is this the best place to put it? Was the class I added appropriate?)
- Added a new `block` in `03-organisms/card-collection/yds-card-collection.twig` that is for adding filters if they exist and to place them below the heading and above the results.
- Updated `atomic/templates/views/views-view--event-list.html.twig` to add the filters into that new block.

Thank you!

joetower commented:
@codechefmarc This is looking good to me! 

I pushed up a couple changes to the [component-library](https://github.com/yalesites-org/component-library-twig/pull/243) and one template addition to the [atomic branch](https://github.com/yalesites-org/atomic/pull/127/commits/bda28c5c0d4447fcc78991c048a5c12b6fa8855c). 

Could you please give my changes a look and make sure everything looks good to you? 

https://pr-276-yalesites-platform.pantheonsite.io/dave-events

codechefmarc commented:
@joetower - thank you for your edits/additions to the component library and atomic! I tested all of that locally and it worked great. I'm going to do a little accessibility additions before merging.

joetower commented:
@codechefmarc Sound great! I'm actually thinking I'm going to change the dropdown icon to the angle-down svg. I'm going to push another fix to the component-library branch and will let you know when my fix has been pushed over there.

joetower commented:
@codechefmarc Okay, I pushed up my fixes to the component-library and rebuilt the mutlidev :partying_face: 
https://pr-276-yalesites-platform.pantheonsite.io/dave-events

nJim commented on package.json:
```diff
@@ -43,5 +43,8 @@
     ]
+  },
+  "dependencies": {
+    "@drupal/autocomplete": "^0.0.2"
```
Remove unused dependency.

nJim reviewed (changes requested):
The combination of any/all terms is working. We're about to introduce a lot more in the filters section, especially with the introduction of news and profiles. Once the other content types catch up, maybe we can crawl through this code and see if there is room to reorganize. Right now it's perfect.
 
Just one change -- something related to my early contribution. The new field is rendering on Event nodes. Can you please remove the category field from the node? We may add it in the future, but right now it is just out of place and unstyled. (https://pr-276-yalesites-platform.pantheonsite.io/admin/structure/types/manage/event/display/default/layout)

nJim reviewed (approved):
I made a quick change to update the display. So this is ready to go. Shipping.
